pkg/detectors/snowflake/snowflake.go

```diff
@@ -0,0 +1,161 @@
+	"net/http"
+	"regexp"
+	"strings"
+	"unicode"
+)
+
+type Scanner struct {
+	client *http.Client
+}
+
+// Ensure the Scanner satisfies the interface at compile time.
+var _ detectors.Detector = (*Scanner)(nil)
+
+var (
+	accountIdentifierPat = regexp.MustCompile(detectors.PrefixRegex([]string{"account"}) + `\b([a-zA-Z]{7}-[0-9a-zA-Z]{7})\b`)
+	usernameExclusionPat = `!@#$%^&*{}:<>,.;?()/\+=\s\n`
+)
+
+const (
+	database                  = "SNOWFLAKE"
+	retrieveAllDatabasesQuery = "SHOW DATABASES"
+)
+
+// Keywords are used for efficiently pre-filtering chunks.
+// Use identifiers in the secret preferably, or the provider name.
+func (s Scanner) Keywords() []string {
+	return []string{"snowflake"}
+}
+
+func meetsSnowflakePasswordRequirements(password string) (string, bool) {
+	var hasLower, hasUpper, hasNumber, minLen bool
+
+	if len(password) < 8 {
+		minLen = false
+	} else {
+		minLen = true
+	}
+
+	for _, char := range password {
+		switch {
+		case unicode.IsLower(char):
+			hasLower = true
+		case unicode.IsUpper(char):
+			hasUpper = true
+		case unicode.IsNumber(char):
+			hasNumber = true
+		}
+
+		if hasLower && hasUpper && hasNumber && minLen {
+			return password, true
+		}
+	}
+
+	return "", false
+}
+
+// FromData will find and optionally verify Snowflake secrets in a given set of bytes.
+func (s Scanner) FromData(ctx context.Context, verify bool, data []byte) (results []detectors.Result, err error) {
+	dataStr := string(data)
+
+	accountMatches := accountIdentifierPat.FindAllStringSubmatch(dataStr, -1)
+
+	usernameRegexState := common.UsernameRegexCheck(usernameExclusionPat)
+	usernameMatches := usernameRegexState.Matches(data)
+
+	passwordRegexState := common.PasswordRegexCheck(" ") // No explicit character exclusions by Snowflake for passwords
+	passwordMatches := passwordRegexState.Matches(data)
+
+	for _, accountMatch := range accountMatches {
+		if len(accountMatch) != 2 {
+			continue
+		}
+		resAccountMatch := strings.TrimSpace(accountMatch[1])
+
+		for _, resUsernameMatch := range usernameMatches {
+
+			for _, resPasswordMatch := range passwordMatches {
+				_, metPasswordRequirements := meetsSnowflakePasswordRequirements(resPasswordMatch)
+
+				if !metPasswordRequirements {
+					continue
+				}
+
+				uri := fmt.Sprintf("%s:%s@%s/%s", resUsernameMatch, resPasswordMatch, resAccountMatch, database)
+
+				s1 := detectors.Result{
+					DetectorType: detectorspb.DetectorType_Snowflake,
+					Raw:          []byte(resPasswordMatch),
+					ExtraData: map[string]string{
+						"account":  resAccountMatch,
+						"username": resUsernameMatch,
+					},
+				}
+
+				if verify {
+					_ = &gosnowflake.Config{
+						Account:  resAccountMatch,
+						User:     resUsernameMatch,
+						Password: resPasswordMatch,
+						Database: database,
+					}
+
+					// Open a connection to Snowflake
+					db, err := sql.Open("snowflake", uri) // Needs the snowflake driver from gosnowflake
```

Comment:
Is there a way to provide a TCP timeout through the driver config or the connection string?

Reply:
The sql.Open function from the database/sql package doesn't provide a direct option to set a timeout for the connection.

Is there a way to disrupt the connection externally in the test?

Reply:
Look at the connection string parameters: https://pkg.go.dev/github.com/snowflakedb/gosnowflake#hdr-Connection_Parameters

Also, I don't think you need to define the config struct, from the looks of it you can just do an unnamed import of the driver and then construct the URI like you're already doing, based on the example at the top of https://pkg.go.dev/github.com/snowflakedb/gosnowflake#section-documentation

Reply:
ah just noticed the loginTimeout param - thanks for the callout.

> loginTimeout: Specifies the timeout, in seconds, for login. The default is 60 seconds. The login request gives up after the timeout length if the HTTP response is success.